packages/protocol/contracts/stability/Exchange.sol

```diff
@@ -113,11 +134,68 @@ contract Exchange is
     updateBucketsIfNecessary
     nonReentrant
     returns (uint256)
+  {
+    return _sell(sellAmount, minBuyAmount, sellGold);
+  }
+
+  /**
+   * @notice Exchanges sellAmount of sellToken in exchange for at least minBuyAmount of buyToken
+   * Requires the sellAmount to have been approved to the exchange
+   * @param sellAmount The amount of sellToken the user is selling to the exchange
+   * @param minBuyAmount The minimum amount of buyToken the user has to receive for this
+   * transaction to succeed
+   * @param sellGold `true` if gold is the sell token
+   * @return The amount of buyToken that was transfered
+   */
+  function _sell(uint256 sellAmount, uint256 minBuyAmount, bool sellGold)
+    private
+    returns (uint256)
   {
     uint256 buyAmount = _getBuyTokenAmount(sellAmount, sellGold);
 
     require(buyAmount >= minBuyAmount, "Calculated buyAmount was less than specified minBuyAmount");
 
+    _exchangeAmounts(sellAmount, buyAmount, sellGold);
+    return buyAmount;
+  }
+
+  /**
+   * @notice Exchanges buyAmount of buyToken in exchange for at most maxSellAmount of sellToken
```

Comment:
```
@notice Exchanges an unspecified amount of one token for a specific amount of another.
@param buyAmount The number of tokens for the exchange to send in return.
@param maxSellAmount The maximum number of tokens to send to the exchange.
@param buyGold True if the exchange is sending CELO to the caller, false otherwise.
@return The number of tokens sent to the exchange.
@dev The caller must first have approved `maxSellAmount` to the exchange.
@dev This function can be frozen via the Freezable interface.
```